### ISSUE TITLE
When I set the batch_size larger,the FLOPs became smaller?

### ISSUE THREAD

**Issue by hanyuchen8145** [Yasir-Ghunaim/RealtimeOCL#2]
When I set the batch_size larger,the FLOPs became smaller,Why?
Please explain it to me.

**Comment by Yasir-Ghunaim** [comment]
Hi @hanyuchen8145,
Can you please provide more details? What values did you use, and which method did you try? It would be great if you could show the script and the results you have.

**Comment by hanyuchen8145** [comment]
[答疑.docx](https://github.com/Yasir-Ghunaim/RealtimeOCL/files/12707580/default.docx)

**Comment by hanyuchen8145** [comment]
[答疑.docx](https://github.com/Yasir-Ghunaim/RealtimeOCL/files/12707581/default.docx)

**Comment by hanyuchen8145** [comment]
@Yasir-Ghunaim
First of all, please forgive me for not replying in time. I'm very sorry.
When I type the following,

python main.py \
--dataset 'cifar10' \
--batch_size 10 \
--lr 0.05 \
--lr_type 'constant' \
--workers 6 \
--method 'ER' \
--seed 123 \
--debug \
--validation \
--profile_flops \
--size_replay_buffer 100

I get 
fwd MACs per GPU=108.7GMACs,
fwd flops of model = fwd flops per GPU*mp_size=219.31G
But when I increase the size of batch_size, like when I increase batch_size to 50, 
for example:

 datasetWithReplay = OnlineDatasetWithReplay(
            dataset=strategy.adapted_dataset,
            mem_size=self.mem_size,
            batch_size=50,
            # batch_size=strategy.train_mb_size,
            gradient_steps=self.gradient_steps,
            batch_delay=self.batch_delay,
            seed=self.seed)

        strategy.dataloader = DataLoader(
            datasetWithReplay,
            num_workers=num_workers,
            batch_size=50,
            # batch_size=strategy.train_mb_size,
            shuffle=shuffle,
            pin_memory=pin_memory,
            **other_dataloader_args

Why do the values fwd MACs per GPU and fwd flops of model = fwd flops per GPU*mp_size decrease? Is that reasonable? 
If it makes sense, can you explain to me why?

**Comment by hanyuchen8145** [comment]
@Yasir-Ghunaim
First of all, please forgive me for not replying in time. I'm very sorry.
When I type the following,

python main.py
--dataset 'cifar10'
--batch_size 10
--lr 0.05
--lr_type 'constant'
--workers 6
--method 'ER'
--seed 123
--debug
--validation
--profile_flops
--size_replay_buffer 100

I get
fwd MACs per GPU=108.7GMACs,
fwd flops of model = fwd flops per GPU*mp_size=219.31G
But when I increase the size of batch_size, like when I increase batch_size to 50,
for example:
datasetWithReplay = OnlineDatasetWithReplay(
dataset=strategy.adapted_dataset,
mem_size=self.mem_size,
batch_size=50,
# batch_size=strategy.train_mb_size,
gradient_steps=self.gradient_steps,
batch_delay=self.batch_delay,
seed=self.seed)

    strategy.dataloader = DataLoader(
        datasetWithReplay,
        num_workers=num_workers,
        batch_size=50,
        # batch_size=strategy.train_mb_size,
        shuffle=shuffle,
        pin_memory=pin_memory,
        **other_dataloader_args
    )
Why do the values fwd MACs per GPU and fwd flops of model = fwd flops per GPU*mp_size decrease? Is that reasonable?
If it makes sense, can you explain to me why?

**Comment by hanyuchen8145** [comment]
@Yasir-Ghunaim
First of all, please forgive me for not replying in time. I'm very sorry.
When I type the following,
python main.py
--dataset 'cifar10'
--batch_size 10
--lr 0.05
--lr_type 'constant'
--workers 6
--method 'ER'
--seed 123
--debug
--validation
--profile_flops
--size_replay_buffer 100
I get
fwd MACs per GPU=108.7GMACs,
fwd flops of model = fwd flops per GPU*mp_size=219.31G.
But when I increase the size of batch_size, like when I increase batch_size to 50,
for example:
datasetWithReplay = OnlineDatasetWithReplay(
    dataset=strategy.adapted_dataset,
    mem_size=self.mem_size,
    batch_size=50,
    # batch_size=strategy.train_mb_size,
    gradient_steps=self.gradient_steps,
    batch_delay=self.batch_delay,
    seed=self.seed)

strategy.dataloader = DataLoader(
    datasetWithReplay,
    num_workers=num_workers,
    batch_size=50,
    # batch_size=strategy.train_mb_size,
    shuffle=shuffle,
    pin_memory=pin_memory,
    **other_dataloader_args
Why do the values fwd MACs per GPU and fwd flops of model = fwd flops per GPU*mp_size decrease? Is that reasonable?
If it makes sense, can you explain to me why?

**Comment by Yasir-Ghunaim** [comment]
@hanyuchen8145 Thanks for the explanation. This is a great finding. 

The primary focus for our FLOPs measurement is indeed to ensure consistency and facilitate comparisons across different methods while maintaining a fixed batch size. The observed changes in FLOPs due to batch size fluctuations are primarily a result of the replay buffer design.

Let's delve deeper into this:
- When we start an iteration with a batch size of 'x,' in the initial step, the buffer is empty. Consequently, the 'OnlineDatasetWithReplay' instance returns 'x' training samples.
- However, as we progress to iteration 2 and beyond, the buffer contains samples. This leads the 'OnlineDatasetWithReplay' instance to return 'x' training samples in addition to 'x' buffer samples. These combined samples construct a dataset of '2x' samples for training.

It's worth noting that in this initial iteration, 'x' buffer samples are effectively skipped during training. Therefore, as the batch size ('x') increases, more buffer samples are skipped, resulting in a reduction in FLOPs usage. Refer to this table for an example:

<img width="600" alt="FLOPs_ER" src="https://github.com/Yasir-Ghunaim/RealtimeOCL/assets/4573867/33ab7843-5eef-4e15-9155-94097f1f7ba9">



As for resolving this issue, if your specific method requires a consistent batch size, there shouldn't be any concerns with the FLOPs calculation since replay is applied uniformly across all methods. However, if your design necessitates varying batch sizes, you can implement a straightforward adjustment. Simply modify line 164 in avalanche_extend/benchmarks/datasets/dataset_wrapper.py from:

```
return sample, target, original_index, \
            test_sample, test_target, test_original_index, test_task_label, task_label
```

to:
```
return sample, target, original_index, \
            test_sample, test_target, test_original_index, test_task_label,
            sample, target, original_index, task_label, task_label
```

This change ensures that the first iteration will return both training and buffer samples, with buffer samples being copied from the training set, thereby maintaining a consistent batch size and FLOPs calculation. Alternatively, you could also enable this behavior for the first iteration only when the profiler_enabled is true by passing it to ReplayOnlinePlugin and then using it internally. This way the training is not affected. 

Note that you may also need to modify other methods that use a different replay implementation such as MIR and GSS.

**Comment by hanyuchen8145** [comment]
@Yasir-Ghunaim 
Ok, I understand, thank you for your explanation!

**Comment by hanyuchen8145** [comment]
Hi@Yasir-Ghunaim,
According to your advice, I simple modify line 164 in avalanche_extend/benchmarks/datasets/dataset_wrapper py
from:
return sample, target, original_index, \
test_sample, test_target, test_original_index, test_task_label, task_label
to:
return sample, target, original_index, \
test_sample, test_target, test_original_index, test_task_label,
sample, target, original_index, task_label, task_label,
Why the results are still larger batch_size and smaller flops may sound strange, but they are. Do you know the reason

**Comment by hanyuchen8145** [comment]
Hi@Yasir-Ghunaim 
According to your advice, I simple modify line 164 in avalanche_extend/benchmarks/datasets/dataset_wrapper py
from:
return sample, target, original_index, \
test_sample, test_target, test_original_index, test_task_label, task_label
to:
return sample, target, original_index, \
test_sample, test_target, test_original_index, test_task_label,
sample, target, original_index, task_label, task_label,
Why the results are still larger batch_size and smaller flops may sound strange, but they are. Do you know the reason

**Comment by hanyuchen8145** [comment]
Hi @Yasir-Ghunaim 
According to your advice, I simple modify line 164 in avalanche_extend/benchmarks/datasets/dataset_wrapper py
from:
return sample, target, original_index, \
test_sample, test_target, test_original_index, test_task_label, task_label
to:
return sample, target, original_index, \
test_sample, test_target, test_original_index, test_task_label,
sample, target, original_index, task_label, task_label,
Why the results are still larger batch_size and smaller flops may sound strange, but they are. Do you know the reason

**Comment by hanyuchen8145** [comment]
> sample, target, original_index, task_label, task_label

Problem solved! Because I made a mistake somewhere else.

**Comment by Yasir-Ghunaim** [comment]
Great, let me know if you have any other issues.
